### PR TITLE
design/complete check process 2020

### PR DESCRIPTION
**Pull request by GuyHarwood**
Revised complete check process.

- dedicated typescript function app `func-complete-check` for all submitted check services
- check-submitted storage queue. replaces check-complete queue
- check receiver. persists submitted check, notifies teacher, forwards for validation.
- check validator. validates submitted check and forwards for marking.
- check marker. marks check and forwards for MI reporting / data warehousing
- strict typing policy.
- tslint implemented and issues resolved
- POC on cosmos table API (under `design/2020/cosmos-sql`)
- exploratory notes `design/2020/design-ideas.md`

**Next steps**

PBIs will be created for the following...

- check receiver: check received notification (sql / redis)
- check validator: object validation currently done dynamically against object property checks.  Leverage TS to force strict typing where possible.
- check marker: implementation
- unit test framework implementation
- check completion: not yet defined.
- deployment: strategy is to implement in parallel until we are forced to break the legacy system.

### CODE REVIEW

File: design/2020/cosmos-sql/index.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  copy.schoolUUID = uuid()
+  copy.checkCode = uuid()
+  copy.createdAt = new Date()
+  return copy
+}
+
+createDatabase()
+  .then(() => createContainer())
+  .then(() => persistCheck(getLargeCheck()))
+  // .then(() => queryCheck(newCheck.checkCode))
```
Comment (GeorgeChatzigiannis) [inline design/2020/cosmos-sql/index.js]: is that temporary commented out?

File: func-complete-check/check-validator/v1.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+const tableService = azureStorageHelper.getPromisifiedAzureTableService()
+import checkSchema from '../messages/complete-check.v1.json'
+
+class V1 {
+  async process (context: Context, validateCheckMessage: ValidateCheckMessageV1) {
+    let receivedCheck = findReceivedCheck(context.bindings.receivedCheckTable)
+    try {
+      detectArchive(receivedCheck)
+      const decompressedString = compressionService.decompress(receivedCheck.archive)
+      const checkData = JSON.parse(decompressedString)
```
Comment (GeorgeChatzigiannis) [inline func-complete-check/check-validator/v1.ts]: do we need to safeguard the parsing at this level?
just curious about the output of the lzstring decompress.

File: func-complete-check/check-validator/v1.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import { Context } from '@azure/functions'
+import { ValidateCheckMessageV1, ReceivedCheck, MarkCheckMessageV1 } from '../typings/message-schemas'
+import Moment from 'moment'
```
Comment (GeorgeChatzigiannis) [inline func-complete-check/check-validator/v1.ts]: Why uppercase on moment?

Comment (GeorgeChatzigiannis) [inline func-complete-check/check-validator/v1.ts]:
```suggestion
import moment from 'moment'
```

File: func-complete-check/check-validator/v1.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    throw new Error(`received check reference was not an array`)
+  }
+  if (receivedCheckRef.length === 0) {
+    throw new Error('received check reference is empty')
+  }
+  return receivedCheckRef[0]
+}
+
+async function updateReceivedCheckWithValidationTimestamp (receivedCheck: ReceivedCheck) {
+  receivedCheck.validatedAt = Moment().toDate()
```
Comment (GeorgeChatzigiannis) [inline func-complete-check/check-validator/v1.ts]:
```suggestion
  receivedCheck.validatedAt = moment().toDate()
```

File: func-complete-check/check-validator/v1.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  if (!message.hasOwnProperty('archive')) {
+    throw new Error(`message is missing 'archive' property`)
+  }
+}
+
+function validateArchive (check: object, context: Context) {
+  // get top level properties of message schema as an array
+  // @ts-ignore
+  const allProperties = Object.getOwnPropertyNames(checkSchema)
+  const requiredProperties = R.without(['version'], allProperties)
```
Comment (GeorgeChatzigiannis) [inline func-complete-check/check-validator/v1.ts]: 👍

File: design/2020/cosmos-sql/index.js (hunk 1 above)
Comment (GuyHarwood) [reply design/2020/cosmos-sql/index.js]: this is just proof of concept code, sketching out ideas and trying out technologies

File: func-complete-check/check-validator/v1.ts (hunk 2 above)
Comment (GuyHarwood) [reply func-complete-check/check-validator/v1.ts]: yes definitely, good point.  this code is not production ready, but thats an important part of it

Comment (GuyHarwood) [reply func-complete-check/check-validator/v1.ts]: i will add this to the ACs of the PBI

File: func-complete-check/check-validator/v1.ts (hunk 3 above)
Comment (GuyHarwood) [reply func-complete-check/check-validator/v1.ts]: that is the type that is exported from the TS definition file

Comment (GuyHarwood) [reply func-complete-check/check-validator/v1.ts]: as above.  Moment is the typescript export

File: func-complete-check/check-validator/v1.ts (hunk 4 above)
Comment (GuyHarwood) [reply func-complete-check/check-validator/v1.ts]: as above

File: func-complete-check/types.d.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+declare module 'azure-storage-helper'
```
Comment (GeorgeChatzigiannis) [inline func-complete-check/types.d.ts]: is that being currently used?

Comment (GuyHarwood) [reply func-complete-check/types.d.ts]: removed, thanks.

File: func-complete-check/check-receiver/index.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import { AzureFunction, Context } from '@azure/functions'
+import * as schemas from '../typings/message-schemas'
+import { performance } from 'perf_hooks'
+const functionName = 'check-receiver'
+import V3 from './v3'
+
+const queueTrigger: AzureFunction = async function (context: Context, completedCheck: schemas.SubmittedCheckMessageV3): Promise<void> {
+  const start = performance.now()
+  const version = completedCheck.version
+  context.log(typeof(version))
```
Comment (jon-shipley) [inline func-complete-check/check-receiver/index.ts]: Looks like incomplete logging

File: func-complete-check/check-receiver/v3.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    const receivedCheckEntity: ReceivedCheck = {
+      PartitionKey: receivedCheck.schoolUUID,
+      RowKey: receivedCheck.checkCode,
+      archive: receivedCheck.archive,
+      checkReceivedAt: Moment().toDate(),
+      checkVersion: +receivedCheck.version
+    }
+    await tableService.insertEntityAsync('receivedCheck', receivedCheckEntity)
+    const message: ValidateCheckMessageV1 = {
+      version: '1',
```
Comment (jon-shipley) [inline func-complete-check/check-receiver/v3.ts]: Should version be an int?

File: func-complete-check/check-validator/index.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+import { performance } from 'perf_hooks'
+const functionName = 'check-validator'
+import V1 from './v1'
+
+const serviceBusQueueTrigger: AzureFunction = async function (context: Context, validateCheckMessage: ValidateCheckMessageV1): Promise<void> {
+  const start = performance.now()
+  const version = validateCheckMessage.version
+  context.log.info(`${functionName}: version:${version} message received for checkCode ${validateCheckMessage.checkCode}`)
+  try {
+    if (version !== '1') {
```
Comment (jon-shipley) [inline func-complete-check/check-validator/index.ts]: Version might be better suited to being an int

File: func-complete-check/check-validator/v1.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  await tableService.replaceEntityAsync('receivedCheck', receivedCheck)
+}
+
+function detectArchive (message: object) {
+  if (!message.hasOwnProperty('archive')) {
+    throw new Error(`message is missing 'archive' property`)
+  }
+}
+
+function validateArchive (check: object, context: Context) {
```
Comment (jon-shipley) [inline func-complete-check/check-validator/v1.ts]: We are going to have to beef this up in some other PRs.

File: func-complete-check/package-lock.json [hunk 11]
```diff
@@ -0,0 +1,2403 @@
+{
+  "requires": true,
```
Comment (jon-shipley) [inline func-complete-check/package-lock.json]: We are using yarn, so there should be a `yarn.lock` but this file should not be needed.

File: func-complete-check/package.json [hunk 12]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "func-complete-check",
+  "version": "0.1.0",
+  "scripts": {
+    "build": "tsc",
+    "build:production": "npm run prestart && npm prune --production",
```
Comment (jon-shipley) [inline func-complete-check/package.json]: s/npm/yarn/

File: functions/completed-checks/function.json [hunk 13]
```diff
@@ -1,5 +1,5 @@
 {
-  "disabled": false,
+  "disabled": true,
```
Comment (jon-shipley) [inline functions/completed-checks/function.json]: This is no longer used.  Disabling a function is now done using env vars. Delete.

File: func-complete-check/check-receiver/index.ts (hunk 7 above)
Comment (GuyHarwood) [reply func-complete-check/check-receiver/index.ts]: this was done to understand why version needed to be a string, will remove.

File: func-complete-check/check-validator/v1.ts (hunk 10 above)
Comment (GuyHarwood) [reply func-complete-check/check-validator/v1.ts]: most definitely.  see the feature for more info.  we can obviously expand this as we see fit

File: func-complete-check/check-receiver/v3.ts (hunk 8 above)
Comment (GuyHarwood) [reply func-complete-check/check-receiver/v3.ts]: that's what i did at first.  but even with the incoming message set to be of type `ValidateCheckMessageV1` it comes in as a string, probably due to the fact that the message body is a string. i will revisit this in the next PBI

File: functions/completed-checks/function.json (hunk 13 above)
Comment (GuyHarwood) [reply functions/completed-checks/function.json]: removed all

File: func-complete-check/package-lock.json (hunk 11 above)
Comment (GuyHarwood) [reply func-complete-check/package-lock.json]: removed

File: func-complete-check/check-validator/index.ts (hunk 9 above)
Comment (GuyHarwood) [reply func-complete-check/check-validator/index.ts]: i tried this, but it doesn't work. (hence the typeof logging above)
when the message is injected into the function via the binding it sets it to a string.
so when it does the check for version, it always fails.
i will leave the interface types as numbers to express our intent, but will keep the comparison as a string to ensure it works.  rubbish 🦆

File: func-complete-check/package.json (hunk 12 above)
Comment (GuyHarwood) [reply func-complete-check/package.json]: updated. prune is obsolete in yarn

**Pull request merged**